### PR TITLE
stash: consolidate in a background task

### LINKED ISSUE #15842: surfaces: stash consolidation is a pathological workload for CockroachDB
One of the causes of #13772.

We tracked down one of the causes of slow DDL in a customer environment today. For internal folks, there is a long Slack thread with the analysis: https://materializeinc.slack.com/archives/CM7ATT65S/p1667362082727089. Here's a tl;dr for external folks:

  * We schedule a stash consolidation after every DDL statement (e.g., `DROP VIEW`, `CREATE VIEW`).
  * Stash consolidation happens on the main coordinator thread. It has lower priority than all other work, but when clients execute DDL In sequence on one connection, the coordinator will almost surely execute.
  * In large customer environments, consolidation can take 5-7s.
  * The cause is the same as #13975. We rewrite all data in the stash collection on every consolidation, which causes a bunch of stale versions to get left behind in Cockroach. After a few hundred or a few thousand DDL statements (easy to do with dbt), enough state has piled up in CockroachDB that scans can take multiple seconds. 

Possible mitigations:

  * Consolidate in a background task that uses a separate CockroachDB connection. Would likely need to figure out how to do incremental consolidation to avoid massive transactions that contend and can never complete.
  * Consolidate less frequently. Ideally we'd only consolidate when a stash collection has doubled in size, to get a logarithmic factor. 
  * Turn down the Cockroach GC TTL to 10m, like we did for persist.

### PR DESCRIPTION
**Pull request by mjibson**
Two major improvements to postgres consolidation.

First: consolidate in a background task using a separate connection. This has the potential to unblock consolidation operations from user requests. However, based on our very slow method of consolidating (deleting and rewriting the whole dataset each consolidation), this operation could take a long time and still block user requests due to transaction contention.

Second: to reduce the overall time consolidation takes and thus reduce transaction contention time, only consolidate keys for which deletions have happened. This leaves lots of data with timestamps before the since, but that's fine. Instead, we only insert a new entry if the diff is non zero, otherwise we delete the two diff entries that now sum to 0. Theoretically we could consolidate and actually produce a new entry, but our use of the stash guarantees that the total diff of an (k,v) pair is either 0 or 1, so consolidation will in practice only delete entries.

This incorporates the first suggestion of the linked issue. Consolidating only negative diffs means the amount of garbage produced goes down significantly, so doing the GC TTL change and "consolidate only when doubled" suggestions probably don't do much now (unverified assumption, should be checked once deployed).

Fixes #15842

```
consolidation           time:   [35.018 ms 36.043 ms 37.285 ms]
                        change: [-17.015% -13.649% -10.048%] (p = 0.00 < 0.05)
                        Performance has improved.
```

### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - n/a

### CODE REVIEW

**Comment by benesch** [comment]
I am out of bandwidth to review this but I am nonetheless very excited about it!

> Consolidating only negative diffs means the amount of garbage produced goes down significantly, so doing the GC TTL change and "consolidate only when doubled" suggestions probably don't do much now (unverified assumption, should be checked once deployed).

Yeah, consolidating only negative diffs seems like it should solve 99% of the problem! Seems like a good reason not to add more complexity like "consolidate only when doubled." But! It still seems to me like it'd be a good idea to set gc.ttlseconds to 10m? It's _super_ simple to do this if we copy/paste what `persist` does:

https://github.com/MaterializeInc/materialize/blob/676634011e49772fc8e9884f96f828e294e288d1/src/persist/src/postgres.rs#L199-L209

@arusahni we're planning to take non-incremental Cockroach backups for all tables, right? So breaking incremental backups on another table won't make your life any harder, will it?

**Comment by mjibson** [comment]
Agree about the ttl. Will do somehow in another PR once cloud signs off.

**Comment by benesch** [comment]
Ah, the full backup work landed! https://github.com/MaterializeInc/cloud/pull/4439/files. Appears to unconditionally do a full backup of everything, so I think we're golden.

File: src/stash/src/postgres.rs [hunk 1]
```diff
@@ -76,12 +78,59 @@ struct PreparedStatements {
...
     seal: Statement,
     compact: Statement,
-    consolidate_delete: Statement,
-    consolidate_consolidate: Statement,
-    consolidate_cleanup: Statement,
     update_many: Statement,
 }
 
+impl PreparedStatements {
+    async fn from(client: &Client) -> Result<Self, StashError> {
```
Comment (jkosh44) [inline src/stash/src/postgres.rs:85]: You might be able to implement the `TryFrom<&Client>` trait here. Though it's not really important.

**Review by jkosh44 — APPROVED** [review]
LGTM

File: src/stash/src/postgres.rs (hunk 1 above)
Comment (mjibson) [reply src/stash/src/postgres.rs:85]: I agree with both sentences and just kinda don't care too much.

**Pull request merged**